akka-cluster-sharding/src/main/scala/akka/cluster/sharding/dynamic/internal/DynamicShardAllocationClientImpl.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.dynamic.internal
+
+import java.util.concurrent.CompletionStage
+
+import akka.Done
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.actor.AddressFromURIString
+import akka.annotation.InternalApi
+import akka.cluster.ddata.DistributedData
+import akka.cluster.ddata.LWWMap
+import akka.cluster.ddata.LWWMapKey
+import akka.cluster.ddata.Replicator.Get
+import akka.cluster.ddata.Replicator.GetFailure
+import akka.cluster.ddata.Replicator.GetSuccess
+import akka.cluster.ddata.Replicator.ReadMajority
+import akka.cluster.ddata.Replicator.Update
+import akka.cluster.ddata.Replicator.UpdateSuccess
+import akka.cluster.ddata.Replicator.UpdateTimeout
+import akka.cluster.ddata.Replicator.WriteMajority
+import akka.cluster.ddata.SelfUniqueAddress
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.dynamic.ClientTimeoutException
+import akka.cluster.sharding.dynamic.DynamicShardAllocationStrategy.ShardLocation
+import akka.cluster.sharding.dynamic.ShardLocations
+import akka.event.Logging
+import akka.util.Timeout
+import akka.util.PrettyDuration._
+import akka.pattern.ask
+
+import scala.concurrent.Future
+import scala.compat.java8.FutureConverters._
+import akka.util.JavaDurationConverters._
+import com.github.ghik.silencer.silent
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final private[dynamic] class DynamicShardAllocationClientImpl(system: ActorSystem, typeName: String)
+    extends akka.cluster.sharding.dynamic.scaladsl.DynamicShardAllocationClient
+    with akka.cluster.sharding.dynamic.javadsl.DynamicShardAllocationClient {
+
+  private val log = Logging(system, classOf[DynamicShardAllocationClientImpl])
+
+  private val replicator: ActorRef = DistributedData(system).replicator
+  private val self: SelfUniqueAddress = DistributedData(system).selfUniqueAddress
+  private val ddataKeys =
+    system.settings.config.getInt("akka.cluster.sharding.dynamic-shard-allocation-strategy.ddata-keys")
+
+  private val DataKeys = (0 until ddataKeys).map(i => LWWMapKey[ShardId, String](s"dynamic-sharding-$typeName-$i"))
+
+  private val timeout =
+    system.settings.config.getDuration("akka.cluster.sharding.dynamic-shard-allocation-strategy.client-timeout").asScala
+  private implicit val askTimeout = Timeout(timeout * 2)
+  private implicit val ec = system.dispatcher
+
+  override def updateShardLocation(shard: ShardId, location: Address): Future[Done] = {
+    val key = DataKeys(shard.hashCode() % ddataKeys)
+    log.debug("updateShardLocation {} {} key {}", shard, location, key)
+    (replicator ? Update(key, WriteMajority(timeout), None) {
```

Comment:
would WriteLocal be enough?

Reply:
Yes I think so. The scenario where the node crashes before it is propagated isn't an issue as for the Kafka use case as the location will be updated from a new node that takes the Kafka partition